pkg/systemd/services/services.jsx

```diff
@@ -266,51 +273,61 @@ class ServicesPageBody extends React.Component {
         cockpit.addEventListener("locationchanged", onLocationChanged);
         onLocationChanged(); // initialize
 
-        /* Start listening to signals for updates - when in the middle of reload mute all signals
+        /* Start listening to signals for updates
+         * - when in the middle of reload mute all signals
          * - We don't need to listen to 'UnitFilesChanged' signal since every time we
-         *   perform some file operation we do call Reload which issues 'Reload' signal
-         * - JobNew is also useless, JobRemoved is enough since it comes in pair with JobNew
-         *   but we are interested to update the state when the operation finished
+         *   perform some file operation we do call Reload which issues 'Reloading' signal
          */
         systemd_client[this.props.owner].subscribe({
             interface: "org.freedesktop.DBus.Properties",
             member: "PropertiesChanged"
         }, (path, iface, signal, args) => {
-            if (this.props.isLoading)
+            if (this.props.isLoading || this.reloading)
+                return;
+
+            // ignore uninteresting interfaces and unit types
+            if (args[0] !== "org.freedesktop.systemd1.Unit" && args[0] !== "org.freedesktop.systemd1.Timer")
+                return;
+            if (!path.endsWith("service") && !path.endsWith("timer") && !path.endsWith("socket"))
                 return;
 
-            if (this.state.unit_by_path[path] &&
-                this.state.unit_by_path[path].Transient &&
-                args[1].ActiveState &&
-                ((args[1].ActiveState.v == 'failed' && this.state.unit_by_path[path].CollectMode == 'inactive-or-failed') ||
-                  args[1].ActiveState.v == 'inactive')) {
-                this.seenPaths.delete(path);
-                const copy_unit_by_path = { ...this.state.unit_by_path };
-                delete copy_unit_by_path[path];
-                this.setState({
-                    unit_by_path: copy_unit_by_path,
-                });
+            const unit_id = this.unit_by_path[path];
+            const unit = this.state.units[unit_id];
+
+            if (unit) {
+                // update existing unit state
+                const newUnit = { ...unit };
+                for (const prop of ["ActiveState", "LoadState", "Description"]) {
+                    if (args[1][prop])
+                        newUnit[prop] = args[1][prop].v;
+                }
+
+                this.setState({ units: { ...this.state.units, [unit_id]: newUnit } });
+                this.processFailedUnits();
             } else {
-                this.updateProperties(args[1], path);
+                // this happens when starting an unloaded unit
+                this.listUnits();
```

Comment:
This surprised me.  The first PropertiesChanged for a just-loaded unit should have everything we need, like the Id, no? Why do we need to list all units?

Reply:
That was my hope, but it's unfortunately not the case. `podman.socket` is normally unloaded for me (as it's disabled), and when I run `systemctl --user start podman.socket` I only get a single PropertiesChanged for the properties related to becoming active, not *all* the properties:

```
XXX PropertiesChanged on unknown unit undefined /org/freedesktop/systemd1/unit/podman_2esocket :·    
Array(3) [ "org.freedesktop.systemd1.Unit", {…}, (2) […] ]    
0: "org.freedesktop.systemd1.Unit"    
1: Object { ActiveState: {…}, FreezerState: {…}, SubState: {…}, … }    
ActiveEnterTimestamp: Object { t: "t", v: 0 }    
ActiveEnterTimestampMonotonic: Object { t: "t", v: 0 }    
ActiveExitTimestamp: Object { t: "t", v: 0 }    
ActiveExitTimestampMonotonic: Object { t: "t", v: 0 }    
ActiveState: Object { t: "s", v: "inactive" }    
AssertResult: Object { t: "b", v: true }    
AssertTimestamp: Object { t: "t", v: 1677673333524198 }    
AssertTimestampMonotonic: Object { t: "t", v: 155255980148 }    
ConditionResult: Object { t: "b", v: true }    
ConditionTimestamp: Object { t: "t", v: 1677673333524182 }    
ConditionTimestampMonotonic: Object { t: "t", v: 155255980131 }    
FreezerState: Object { t: "s", v: "running" }    
InactiveEnterTimestamp: Object { t: "t", v: 0 }    
InactiveEnterTimestampMonotonic: Object { t: "t", v: 0 }    
InactiveExitTimestamp: Object { t: "t", v: 0 }    
InactiveExitTimestampMonotonic: Object { t: "t", v: 0 }    
InvocationID: Object { t: "ay", v: "6X0LoHmbRxCMYzAaDi1AFg==" }    
Job: Object { t: "(uo)", v: (2) […] }    
StateChangeTimestamp: Object { t: "t", v: 0 }    
StateChangeTimestampMonotonic: Object { t: "t", v: 0 }    
SubState: Object { t: "s", v: "dead" }    
```
I'll add a comment to point this out. We of course could also do a GetAll call. Anyway, I don't like the listUnits() call here, it is too heavy and also causes the "Loading" spinner to re-appear (as you pointed out). I even had a "FIXME" comment on this once, but that apparently got lost somewhere. I.e. this is one of the reasons why this is still "draft", and I like to improve this.

Reply:
I experimented with this a bit; we could listen for `UnitNew`, which comes before the PropertiesChanged, but it's more expensive -- it does not give much information, and needs *two* D-Bus calls to fetch the unit runtime and the unit file properties. But the same is true with doing a GetAll() call, as that  still does not have UnitFilePath and UnitFileState. Also, this does not really hurt much. I'll leave this as future possible optimization.

Reply:
That's a bit disappointing, no? listUnits is certainly more expensive than GetAll?

Reply:
Yeah, perhaps, but it's not so clear-cut. Anything that reintroduces listening to UnitNew or JobNew is off the table IMHO, as these were exactly the things that caused these hundreds of D-Bus messages. But doing two GetAll() calls to retrieve the properties sounds fine.

I'm willing to give this another try; just today that effing location bug sucked away all my energy. That's still breaking most of the test runs. Once this is figured out, working on further cleanups like this one is more fun.

Reply:
I reworked the state management again. It avoids this expensive call, and it's much cleaner now. That should also robustify the property updates and [flakes like this](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18422-20230303-201537-c95d73be-fedora-37-devel/log.html#339).

Reply:
Rebased, and keeping the most recent rewrite, which doesn't do this expensive operation any more.